hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1640 @@
+//!
+//!  Set the pointer to [`HashArea`] via [`Icm::set_hash_addr()`]
+//!
+//!  **See note about memory safety above**
+//!
+//! ### Hash calculation
+//!
+//! Assuming general setup is already done, modify the [`RegionConfiguration`]
+//! which is part of the [`MainRegionDesc`]:
+//!
+//! * [`RegionConfiguration::set_rhien()`] to `false` to allow interrupts when
+//!   calculation is done
+//! * [`RegionConfiguration::set_eom()`] to `true` only for the last region
+//!
+//! Change [`RegionAddress`] to point to the object to SHA-sum with
+//! [`MainRegionDesc<RegionNumT>::set_region_address()`]
+//!
+//! ### Memory monitoring
+//!
+//! [`HashArea`] needs to contain the expected SHA-sums of the data to
+//! monitor, [`Icm::set_ascd()`] is provided to help with creating this data.
+//! Alternatively do it manually and then change mode, or prepopulate the
+//! [`HashArea`] with SHA-sums.
+//!
+//! Assuming general setup is already done, modify the [`RegionConfiguration`]
+//! which is part of the [`MainRegionDesc`]:
+//!
+//! * [`RegionConfiguration::set_dmien()`] to `false` to allow interrupts if
+//!   mismatch occurs
+//! * [`RegionConfiguration::set_cdwbn()`] to `true` to change to monitor mode
+//! * [`RegionConfiguration::set_wrap()`] to `true` only for the last region if
+//!   continuous monitoring is desired
+//!
+//! ## Examples
+//!
+//! ### Calculate SHA1, SHA224 and SHA256 sums, then switch to memory monitor
+//!
+//! 4 memory regions, SHA1 in region0 and region1.
+//! Region2 uses SHA224 and region3 SHA256.
+//!
+//! This only covers the setup part, to achieve the functionality of first
+//! computing the SHA-sums and then do region monitoring handling of
+//! interrupts and changing mode is required.
+//!
+//! ```no_run
+//! # use atsamd_hal::{pac::Peripherals, icm::*};
+//!
+//! // SHA Test data
+//! static MESSAGE_REF0: [u32; 16] = [
+//!     0x11111111, 0x22222222, 0x33333333, 0x44444444, 0x55555555, 0x66666666, 0x77777777, 0x88888888,
+//!     0x99999999, 0xaaaaaaaa, 0xbbbbbbbb, 0xcccccccc, 0xdddddddd, 0xeeeeeeee, 0xffffffff, 0x00000000,
+//! ];
+//!
+//! static MESSAGE_REF1: [u32; 16] = [
+//!     0x80636261, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+//!     0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x18000000,
+//! ];
+//!
+//! // Expected SHA1 sum result
+//! static MESSAGE_SHA1_RES: [u32; 8] = [
+//!     0x363e99a9, 0x6a810647, 0x71253eba, 0x6cc25078, 0x9dd8d09c, 0x00000000, 0x00000000, 0x00000000,
+//! ];
+//!
+//! static MESSAGE_SHA224_RES: [u32; 8] = [
+//!     0x227d0923, 0x22d80534, 0x77a44286, 0xb355a2bd, 0xe4bcad2a, 0xf7b3a0bd, 0xa79d6ce3, 0x00000000,
+//! ];
+//! static MESSAGE_SHA256_RES: [u32; 8] = [
+//!     0xbf1678ba, 0xeacf018f, 0xde404141, 0x2322ae5d, 0xa36103b0, 0x9c7a1796, 0x61ff10b4, 0xad1500f2,
+//! ];
+//! static mut HASH: HashArea = HashArea::default();
+//! static mut ICM_REGION_DESC: Regions = Regions::default();
```

Comment:
Would it be worth using [`cortex_m::singleton`](https://docs.rs/cortex-m/0.7.3/cortex_m/macro.singleton.html) for these `static mut`s? I think it's a bit safer. I've heard `static mut` values are notoriously UB-prone. If nothing else, it might set a good example for people referring to your documentation.

Reply:
Sounds like a good idea. I weren't able to use it in the RTIC demo-app, but I added it to documentation for all other use-cases, see https://github.com/atsamd-rs/atsamd/pull/480/commits/2675d6211566c7ef8b3e5c0dd21e221d5d06a014